File: workers/src/main/scala/com/maxmouchet/vamk/timetables/updater/DBClient.scala
package com.maxmouchet.vamk.timetables.updater

import org.postgresql.ds.PGPoolingDataSource
import java.sql.PreparedStatement
import java.util.Locale
import java.sql.Timestamp
import com.maxmouchet.vamk.timetables.parser.Schedule

class DBClient(source: PGPoolingDataSource) {

  val conn = source.getConnection()

  def insertCourse(name: String) = {
    val st = conn.prepareStatement("INSERT INTO courses(name) VALUES(?)")
    st.setString(1, name)
    st.executeUpdate
    st.close
  }

  def insertSchedule(schedule: Schedule) = {
    var courseId = findCourseId(schedule.courseName)
    if (courseId == -1) {
      insertCourse(schedule.courseName)
      courseId = findCourseId(schedule.courseName)
    }

    val st = conn.prepareStatement("INSERT INTO schedules(course_id, room, professor, start_time, end_time, \"group\") VALUES(?, ?, ?, ?, ?, ?)")
    st.setInt(1, courseId)
    st.setString(2, schedule.room)
    st.setString(3, schedule.professor)
    st.setTimestamp(4, new Timestamp(schedule.startDate.getMillis))
    st.setTimestamp(5, new Timestamp(schedule.endDate.getMillis))
    st.setString(6, schedule.group)

    st.executeUpdate
    st.close
  }

  def dropAllCourses = {
    val st = conn.prepareStatement("DELETE FROM courses")
    st.executeUpdate
    st.close
  }

  def dropAllSchedules = {
    val st = conn.prepareStatement("DELETE FROM schedules")
    st.executeUpdate
    st.close
  }

  def findCourseId(name: String) = {
    var result = -1

    val st = conn.prepareStatement("SELECT id FROM courses WHERE name = ?")
    st.setString(1, name)

    try {
      val results = st.executeQuery
      results.next
      result = results.getInt(1)
    } catch {
      case _: Throwable => //
    } finally {
      st.close
    }

    result
  }

  def findSchedule(name: String) = {
    var result = -1

    val st = conn.prepareStatement("SELECT id FROM schedules WHERE name = ?")
    st.setString(1, name)

    try {
      val results = st.executeQuery
      results.next
      result = results.getInt(1)
    } catch {
      case _: Throwable => //
    } finally {
      st.close
    }

    result
  }

}

File: api_old/db_client.rb
require 'pg'

class DBClient

  def initialize(connection_hash)
    @conn = PG.connect(connection_hash)
  end

  def get_groups(course = nil)
    params = []

    params << course if course

    query = 'SELECT schedules.group FROM schedules GROUP BY schedules.group'
    query = 'SELECT schedules.group FROM schedules WHERE schedules.course_id = $1 GROUP BY schedules.group' if course

    @conn.prepare('stmt', query)
    @conn.exec_prepared('stmt', params) do |results|
      groups = []

      results.each do |group|
        groups << group.values_at('group').first.strip
      end

      groups
    end
  end

  def get_courses
    courses = []

    query = 'SELECT * FROM courses'

    @conn.exec(query) do |results|
      results.each do |course|
        courses << {
          id: course.values_at('id').first,
          name: course.values_at('name').first.strip
        }
      end
    end

    courses
  end

  def get_course(id, group = nil)
    params = []

    params << id
    params << group if group

    query = 'SELECT DISTINCT schedules.id, schedules.room, schedules.professor,
                           schedules.start_time, schedules.end_time, schedules.group, courses.name
                           FROM schedules
                           INNER JOIN courses ON schedules.course_id = courses.id
                           WHERE courses.id = $1'

    query += ' AND schedules.group = $2' if group

    @conn.prepare('stmt', query)
    @conn.exec_prepared('stmt', params) do |results|
      schedules = []

      results.each do |schedule|
        schedules << {
            id: schedule.values_at('id').first,
            text: "#{ schedule.values_at('name').first.strip }\n#{ schedule.values_at('room').first.strip }",
            start_date: DateTime.parse(schedule.values_at('start_time').first).strftime('%m/%d/%Y %H:%M'),
            end_date: DateTime.parse(schedule.values_at('end_time').first).strftime('%m/%d/%Y %H:%M')
          }
      end

      schedules
    end
  end

end
